pkg/queue/manager.go

```diff
@@ -762,3 +779,38 @@ func (m *Manager) DeleteSnapshot(cq *kueue.ClusterQueue) {
 	defer m.snapshotsMutex.Unlock()
 	delete(m.snapshots, kueue.ClusterQueueReference(cq.Name))
 }
+
+func (m *Manager) DeleteSecondPass(w *kueue.Workload) {
+	m.Lock()
+	defer m.Unlock()
+	m.DeleteSecondPassWithoutLock(w)
+}
+
+func (m *Manager) DeleteSecondPassWithoutLock(w *kueue.Workload) {
+	m.secondPassQueue.deleteByKey(workload.Key(w))
+}
+
+func (m *Manager) QueueSecondPassIfNeeded(ctx context.Context, w *kueue.Workload) bool {
+	if workload.NeedsSecondPass(w) {
+		m.clock.AfterFunc(time.Second, func() {
+			m.queueSecondPass(ctx, w)
+		})
+		return true
+	}
+	return false
+}
+
+func (m *Manager) queueSecondPass(ctx context.Context, w *kueue.Workload) {
+	m.Lock()
+	defer m.Unlock()
+	m.queueSecondPassWithoutLock(ctx, w)
+}
```

Comment:
Ditto: Do we need to lock manager if we already have lock on second_pass_queue underneath?

Reply:
I think the manager lock is needed, if we can call broadcast. Consider the manager code:

```golang
func (m *Manager) Heads(ctx context.Context) []workload.Info {
	m.Lock()
	defer m.Unlock()
	log := ctrl.LoggerFrom(ctx)
	for {
		workloads := m.heads()
		log.V(2).Info("Obtained ClusterQueue heads", "count", len(workloads))
		if len(workloads) != 0 {
			return workloads
		}
		select {
		case <-ctx.Done():
			return nil
		default:
			m.cond.Wait()
		}
	}
}
```
consider the sequence of calls:
1. `m.heads()` returns empty
2. another `queueSecondPass` does not take m.Lock and adds something to secondPassQueue, sending Broadcast
3. `m.cond.Wait()` is called, and we wait indefinitely on Broadcast, even though there is an item waiting in the queue